sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,156 @@
+            {
+                if (HasValue)
+                {
+                    return _value;
+                }
+                else if (_requestFailedException != null)
+                {
+                    throw _requestFailedException;
+                }
+                else
+                {
+                    throw new InvalidOperationException("The operation has not completed yet.");
+                }
+            }
+        }
+
+        public Response GetRawResponse() => _rawResponse;
+
+        public async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken) =>
+            await UpdateStatusAsync(async: true, cancellationToken).ConfigureAwait(false);
+
+        public Response UpdateStatus(CancellationToken cancellationToken) =>
+            UpdateStatusAsync(async: false, cancellationToken).EnsureCompleted();
+
+        public async ValueTask<Response<TValue>> WaitForCompletionAsync(CancellationToken cancellationToken) =>
+            await WaitForCompletionAsync(_defaultPollingInterval, cancellationToken).ConfigureAwait(false);
+
+        public async ValueTask<Response<TValue>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, GetRawResponse());
+                }
+
+                await Task.Delay(pollingInterval, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            if (HasCompleted)
+            {
+                return GetRawResponse();
+            }
+
+            var scopeName = async
+                ? $"{_operationTypeName}.UpdateStatusAsync"
+                : $"{_operationTypeName}.UpdateStatus";
+
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope(scopeName);
+            scope.Start();
+
+            try
+            {
+                var response = async
+                    ? await _getResponseAsync(cancellationToken).ConfigureAwait(false)
+                    : _getResponse(cancellationToken);
+
+                _rawResponse = response.GetRawResponse();
+
+                var status = _getStatus(response);
+
+                if (status == OperationInternalResponseStatus.Succeeded)
+                {
+                    _value = _parseResponse(response);
+                    HasValue = true;
+                    HasCompleted = true;
+                }
+                else if (status == OperationInternalResponseStatus.Failed)
+                {
+                    _requestFailedException = _getFailure(response);
+                    HasCompleted = true;
+
+                    throw _requestFailedException;
+                }
```

Comment:
We're using a new enum (`OperationInternalResponseStatus`) to decide whether operation has failed or not, which adds a layer of complexity to the approach.

We could maybe have a single method `TryParseResponse`:
- Returns true and parses response if succeeded.
- Returns false if pending.
- Throws if failure.